src/decoder/cuda-lattice-decoder.cu

```diff
@@ -0,0 +1,1652 @@
+                       config.max_tokens, config.max_arcs);
+
+  lat_arcs_buf_.Allocate(config.max_arcs, NULL, NULL, NULL,
+                         lattice_pruner_.GetDeviceArcsBpr());
+  bytes_cuda_malloc += lat_arcs_buf_.GetCudaMallocBytes();
+
+  for (int32 j = 0; j < LAT_BUF_SIZE; j++) {
+    lat_toks_bufs_[j].Allocate(config.max_tokens_per_frame);
+    bytes_cuda_malloc += lat_toks_bufs_[j].GetCudaMallocBytes();
+  }
+
+  // In each frame, we save the token 
+  // information in an array whose size is the number of arcs. This
+  // ensures there are no write conflicts between threads since each
+  // arc can be accessed at most once in each frame. It's a temp solution
+  cudaMalloc((void**)&token_per_arc_d, sizeof(Token)*fst.NumArcs()); 
+  cudaMalloc((void**)&token_per_arc_update_d,
+             sizeof(int32)*fst.NumArcs()); //temp solution
+  cudaMemset(token_per_arc_update_d, 0,
+             sizeof(int32)*fst.NumArcs()); //temp solution
+  bytes_cuda_malloc += (sizeof(Token) + sizeof(int32)) * (fst.NumArcs());
+
+  num_frames_decoded_ = 0;
+  UpdateTokPointersByFrame(num_frames_decoded_);
+
+  cudaStreamSynchronize(stream_comp);
+  cudaStreamSynchronize(stream_lat[0]);
+  cudaStreamSynchronize(cudaStreamPerThread);
+  //sgemm requires shared memory and we don't want cache config changing.  
+  // So set a device wide cache config.
+  cudaDeviceSetCacheConfig(cudaFuncCachePreferEqual);
+}
+
+CudaLatticeDecoder::~CudaLatticeDecoder() {
+  KALDI_VLOG(1) << "CUDA LatticeDecoder DESTRUCTOR\n";
+
+  for (int32 j = 0; j < LAT_BUF_SIZE; j++) lat_toks_bufs_[j].Free();
+  lat_arcs_buf_.Free(true);
+  lattice_pruner_.Free();
+  token_allocator_.Finalize();
+
+  cudaFreeHost(loglikelihoods_h);
+  cudaFreeHost(loglikelihoods_old_h);
+  cudaFree(loglikelihoods_d);
+  cudaFree(loglikelihoods_old_d);
+  cudaFree(current_tokens_lookup_d);
+
+  cudaFree(pe_idx_d);
+  cudaFree(agg_idx_d);
+  cudaFree(ne_idx_d);
+  cudaFree(ne_queue_d);
+  cudaFree(fb_idx_d);
+  cudaFree(barrier_d);
+
+  cudaFree(cutoff_d);
+  cudaFree(modified_d);
+
+  cudaFree(token_per_arc_d);
+  cudaFree(token_per_arc_update_d);
+
+  cudaEventDestroy(event_pt);
+  cudaEventDestroy(event_ll);
+  cudaStreamDestroy(stream_comp);
+  for (int32 i = 0; i < LAT_BUF_SIZE; i++)
+    cudaStreamDestroy(stream_lat[i]);
+  cudaStreamDestroy(stream_ll);
+}
+
+void CudaLatticeDecoder::ComputeLogLikelihoods(DecodableInterface *decodable) {
+  PUSH_RANGE("ComputeLogLikelihoods", 3)
+  int32 frame = num_frames_decoded_;
+  // finish decoding this frame, it has been ensured outside
+  // cudaStreamSynchronize(stream_comp); 
+
+  //double buffering so we don't overwrite loglikelihoods_h before it is copied down  
+  std::swap(loglikelihoods_h,
+            loglikelihoods_old_h); 
+  std::swap(loglikelihoods_d, loglikelihoods_old_d); 
+
+  decodable->ComputeLogLikelihoods(loglikelihoods_h, frame, fst_.max_ilabel + 1);
+
+  //copying in another stream to overlap transfer with compute
```

Comment:
I see a lot of this going on. You can use a simple regular expression to find all the //[a-zA-z] in the code and add apaces